### PR TITLE
Better contact removal dialog

### PR DESCRIPTION
**Pull request by ghost**
This dialog solves all confusion that arose from current verification method and is accident-proof. Current method asks if user also wants to remove chat history. However confused user may not carefully read message and click "no" because he does not want to remve contact. However in this case contact is in fact removed but history is preserved. It is also open to possiblity that key smashing deletes contact by accident. This is very inconvenient because tox ids are long and hard to memorize. If someone removes contact by accident then he would need to find id of that contact in order to contact him/her. Sometimes this may be difficult.

New contact removal confirmation dialog prompts use to enter "YES" in the text boxin order to enable "OK" button so removal can be accepted. It also has checkbox for history removal. Dialog will work with translations. Russian user would need to enter "Да" in order to accept removal.

rmc

Please test if dialog placement is alright. On my machine its weird, not centered on parent widget. Any advice how to solve this would be greatly appreciated. So far i tried centering it with `move()` but no luck no matter what i do.

### CODE REVIEW

**Comment by DanTheBritish** [comment]
The red is hard for me to read on that background, possible for it to be 1 or two shades darker? (Opinions needed)

**Comment by ghost** [comment]
REMOVE or *remove* would probably be better, no extra colours would be necessary.

**Comment by DanTheBritish** [comment]
What about just regular bold, rather than changing the styling of the letters?

**Comment by ghost** [comment]
The contact name is already in bold, otherwise it looks strange.

**Comment by suhr** [comment]
Too much protecting. Also, red `remove` is unreadable and takes too much attention.
But “Remove history” as a checkbox is a good idea.

**Comment by agilob** [comment]
What's the difference between OK and YES?
I think is should be two buttons Remove | Cancel

**Comment by ProMcTagonist** [comment]
@agilob YES is a text box.

**Comment by PKEv** [comment]
I think flag "also remove chat history" must be setup by default

**Comment by PKEv** [comment]
Typing "yes" is not quite intuitive, can add a label specifying what to do...

**Comment by ghost** [comment]
For me "remove" is not hard to read at all. Besides "that background" is my desktop theme, dont think it would be everywhere like that ;) This dialog follows Qt theme that is set.

Regarding textbox - i would preffer less labels with explanations but i probably should add tooltips/help text (used with question mark button near maximize button). Besides its just like web2.0. Never seen textbox with gray text saying "e-mail" which disappears when you start typing in email?

**Comment by suhr** [comment]
That red color (and any emphasis of the “remove” word) is not useful at all. Better to have “Remove” instead of “OK” as the confirm button.
Also, people who runs bots won't like “YES” typing. Can this be optional?

**Comment by ghost** [comment]
Ok i guess it makes sense.

Err what bots? since when qTox is for bots?

**Comment by suhr** [comment]
> since when qTox is for bots?

Since time people run bots and add them temporary as contacts in qTox.

**Comment by ghost** [comment]
I dont get it..

**Comment by zetok** [comment]
> For me "remove" is not hard to read at all

Why do you consider your own case as the only valid one?

**Comment by suhr** [comment]
> I dont get it..
1. Run bot
2. Add it to your qTox
3. Use the bot
4. Halt the bot
5. Remove the bot from your qTox

**Comment by ghost** [comment]
And you are saying this is done so often that this safeguard makes it VERY inconvenient to use qTox?

**Comment by suhr** [comment]
I won't say “VERY”, but are you sure the safeguard is THAT useful?

**Comment by ghost** [comment]
We already have some people who managed to do some keyboard-smash-contact-removal by accident. Current confirmation is certainly too lacking. This is more of a problem with tox because we can not memorize IDs of our friends and if one is accidentally removed then its cumbersome to re-add him/her. If you have better ideas how to make keyboard-smash-proof contact removal confirmation im all ears.

Regarding bots.. I am not sure its worth to add yet another option for 0.001% of total qTox users (if we are lucky). I think everyone would benefit more from universally good design rather than option.

One thing i can think of is confirmation slider, on one end being "No", on the other "Yes". But i have not seen such thing anywhere and i am not sure it would intuitive for everyone.

**Comment by ProMcTagonist** [comment]
There's another issue open right now for having a list of recently removed friends you can easily re-add. Something to keep in mind when considering this removal protection at the cost of convenience.

**Comment by ghost** [comment]
I would argue that such list is overengineering. If we remove contact means we do not want it. Would it not make more sense to prevent accidents instead of adding yet another rarely used feature? Besides how often do we actually remove contacts to call typing in 3 letter word inconvenient? I personally have not removed contact at least in a year (counting jabber too mind you). People close to me have not done it in months. To call it inconvenient i think we need better arguments.

@tux3 @dubslow care to state your opinions?

**Comment by ghost** [comment]
Renamed contact -> friend to match rest of tox code (filenames, class names). Removed bold red "remove" and changed button from "OK" to "Remove".

rmc

**Comment by zetok** [comment]
I applied this patch on top of patch from #1938, and as result:

```
In file included from src/widget/widget.cpp:51:0:
src/widget/tool/removefrienddialog.h:6:35: fatal error: ui_removefrienddialog.h: No such file or directory
 #include "ui_removefrienddialog.h"
                                   ^
compilation terminated.
Makefile:2503: recipe for target 'widget.o' failed
```

Same happens without #1938 applied.

**Comment by ghost** [comment]
I did evil thing of overwriting this branch so not to trash commit history with these renames that should not be there in the first place. I bet if you delete that branch and check out again it would work. (because it works for me and i double-checked that no files are missing from commit)

P.S. please report how dialog placement works on your system. On mine its funky and im not sure why. But that surely needs solving.

**Comment by zetok** [comment]
> I bet if you delete that branch and check out again it would work. (because it works for me and i double-checked that no files are missing from commit)

nope, I didn't checkout anything.

What I did, is simple `wget` patch from this PR.

And no, after getting new patch, it still doesn't work on top of master.

**Comment by ghost** [comment]
Oh sorry, my bad. I had qTox.pro customized and i did not add stage it for commit because of those custom changes, completely forgetting about new aditions required for this PR. Now it should be good.

**Comment by zetok** [comment]
@r-ku your commit message has gone.. MIA?

**Comment by ghost** [comment]
Thank you for spotting this! Must be some ctrl-v accident. Fixed now.

**Comment by suhr** [comment]
@tux3 @dubslow Merge it?

**Comment by suhr** [comment]
cast @agilob

File: src/widget/tool/removefrienddialog.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    setWindowFlags(windowFlags() | Qt::Tool);
+    ui.setupUi(this);
+    ui.label->setText(ui.label->text().replace("&lt;name&gt;", f->getDisplayedName()));
+    auto removeButton = ui.buttonBox->button(QDialogButtonBox::Ok);
+    removeButton->setEnabled(false);
+    removeButton->setText(tr("Remove"));
+    connect(ui.yes, SIGNAL(textChanged(QString)), this, SLOT(onTextChanged(QString)));
+    connect(ui.buttonBox, SIGNAL(accepted()), this, SLOT(onAccepted()));
+    connect(ui.buttonBox, SIGNAL(rejected()), this, SLOT(close()));
```
Comment (agilob) [inline src/widget/tool/removefrienddialog.cpp]: Could you change it to new Qt5 SIGNAL/SLOT syntax?
http://wiki.qt.io/New_Signal_Slot_Syntax

**Comment by agilob** [comment]
Oh how nice! I left a comment about old singal/slot syntax, I'll merge it when it's fixed!

**Comment by ghost** [comment]
Did as asked. Also figured out that weird dialog placement, now its ok.

**Comment by antis81** [comment]
@agilob Think you want to merge this. Looks good to me.

**Pull request merged**